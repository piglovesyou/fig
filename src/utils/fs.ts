// For Node v12
import { promises } from 'fs';

export const { readFile, writeFile } = promises;
