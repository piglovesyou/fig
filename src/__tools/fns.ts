import { join } from 'path';
import { renderToStaticMarkup } from 'react-dom/server';
import { readFile } from '../utils/fs';

export function renderInHtml(element: JSX.Element) {
  return `<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
<style>
  
body {
  margin: 0;
  padding: 0;
  font-family: sans-serif;
  position: absolute;
  width: 100vw;
  min-height: 100vh;
}

body > * {
  overflow: hidden;
  min-width: 100vw;
  min-height: 100vh;
}

/*.vector svg {*/
/*  width: 100%;*/
/*  height: 100%;*/
/*  position: absolute;*/
/*}*/

/*a {*/
/*  cursor: pointer;*/
/*}*/

</style>
</head>
<body>${renderToStaticMarkup(element)}</body>
</html>`;
}

export async function readJson(...paths: string[]) {
  const fullPath = join(...paths);
  return JSON.parse(await readFile(fullPath, 'utf-8'));
}
