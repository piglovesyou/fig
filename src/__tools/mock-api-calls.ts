import { readFile } from 'fs/promises';
import { join } from 'path';
import * as api from '../core/api';
import { RequestImagesResponse } from '../core/api';

async function read(rel: string) {
  return JSON.parse(await readFile(join(__dirname, rel), 'utf-8'));
}

jest.spyOn(api, 'requestFile').mockImplementation(() => {
  return read('./__fixtures/request-file-response.json');
});

jest.spyOn(api, 'requestVectors').mockImplementation(() => {
  return read('./__fixtures/request-images-response.json');
});

jest.spyOn(api, 'requestImages').mockImplementation(() => {
  return Promise.resolve({
    error: false,
    status: 200,
    meta: {
      images: {},
    },
  } as RequestImagesResponse);
});
