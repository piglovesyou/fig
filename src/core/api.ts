import chunk from 'lodash.chunk';
import fetch from 'node-fetch';
import pMap from 'p-map';
import { FigmaFile } from '../types/fig';
import { updateLog } from './print';

const baseUrl = 'https://api.figma.com';

// To avoid "Error: 414: Request-URI Too Large"
const MAX_VECTOR_REQUEST_COUNT = 200;

// I think I should say hi before my account gets frozen
const GREET_FIGMA_DEVELOPER = `Hi. Sorry for high load. Contact me on github.com/piglovesyou/fig`;

export function makeHeader(token?: string) {
  const h: Record<string, string> = { 'X-Greeting': GREET_FIGMA_DEVELOPER };
  if (token) h['X-Figma-Token'] = token;
  return h;
}

export async function requestVectors(
  fileKey: string,
  vectorList: string[],
  token: string
) {
  if (!vectorList.length) throw new Error('vectorList is empty');

  let result: Record<string, string> | null = null;

  await pMap(
    chunk(vectorList, MAX_VECTOR_REQUEST_COUNT),
    async (ids, i) => {
      updateLog(
        `Fetching vector info ${i * MAX_VECTOR_REQUEST_COUNT}/${
          vectorList.length
        } starting..`
      );
      const idValue = encodeURIComponent(ids.join(','));
      const {
        err,
        images,
      }: {
        err: any;
        images?: Record<string, string>;
      } = await callApi(
        'GET',
        `/v1/images/${fileKey}?ids=${idValue}&format=svg`,
        token
      );
      if (err) throw new Error(JSON.stringify(err));
      if (images) result = Object.assign(result || {}, images);
    },
    { concurrency: 100 }
  );

  return result as unknown as Record<string, string>;
}

export interface RequestImagesResponse {
  error: boolean;
  status: number;
  meta: {
    images: Record<string, string>;
  };
}

export async function requestImages(
  fileKey: string,
  token: string
): Promise<RequestImagesResponse> {
  return await callApi('GET', `/v1/files/${fileKey}/images`, token);
}

/**
 * Get an AST for a Figma File.
 */
export async function requestFile(
  fileKey: string,
  token: string
): Promise<FigmaFile> {
  return await callApi('GET', `/v1/files/${fileKey}`, token);
}

export async function callApi(
  method: 'GET' | 'POST',
  endpoint: string,
  token: string
) {
  const headers = makeHeader(token);
  const resp = await fetch(`${baseUrl}${endpoint}`, { method, headers });
  if (resp.status < 200 || 300 <= resp.status)
    throw new Error(`${resp.status}: ${resp.statusText}\n${await resp.text()}`);
  return await resp.json();
}
