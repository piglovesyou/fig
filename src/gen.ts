import { writeFile } from 'fs/promises';
import { requestFile } from './api';
import { makeGenContext } from './make-gen-context';
import { processCanvas } from './process-canvas';
import { Canvas, FigmaFile } from './types';

export async function gen(outDir: string, fileKey: string) {
  const figmaFile: FigmaFile = await requestFile(fileKey);

  // TODO: for dev
  await writeFile(
    `${__dirname}/__fixtures/${fileKey}.file.json`,
    JSON.stringify(figmaFile, undefined, 2)
  );
  const genContext = await makeGenContext(figmaFile, fileKey, outDir);

  for (const canvas of figmaFile.document.children as Canvas[])
    await processCanvas(canvas, genContext);
}
