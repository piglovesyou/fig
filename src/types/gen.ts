import { FigConfig } from '../core/config';
import { ComponentsMap } from '../gen/components-map';
import { ComposableNode } from './ast';
import { FigPlugin } from './strategy';

export type ComponentInfo = {
  name: string;
  node: ComposableNode;
};

export type GenContext = {
  componentsMap: ComponentsMap;
  imagesMap: Map<string, string>;
  vectorsMap: Map<string, string>;
  config: FigConfig;
  cwd: string;
  libDir: string;

  strategy?: FigPlugin;

  baseFullDir: string;
  componentsFullDir: string;
  pagesFullDir: string;
  htmlFullDir: string;
  imagesFullDir: string;
  // mode: 'generate' | 'edit'
};
