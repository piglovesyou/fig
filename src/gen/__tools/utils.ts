import generate from '@babel/generator';
import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';
import { format } from 'prettier';
import { createConfig } from '../../core/config';
import { ComposableNode, FigmaFile } from '../../types/fig';
import { readFile } from '../../utils/fs';
import { makeGenContext } from '../make-gen-context';

export async function makeCreateComponentArgs(
  fileJsonFullPath: string,
  fileKey: string,
  baseDir: string
) {
  const figmaFileFixture = JSON.parse(
    await readFile(fileJsonFullPath, 'utf-8')
  );
  const node: ComposableNode =
    figmaFileFixture.document.children[0].children[0];
  expect(node.name).toBe('My Frame');

  const config = await createConfig({
    fileKeys: [fileKey],
    baseDir,
    plugins: ['react'],
  });
  const genContext = await makeGenContext(
    figmaFileFixture as FigmaFile,
    fileKey,
    config,
    baseDir
  );

  const { componentsMap } = genContext;
  const componentInfo = componentsMap.get('1:2');
  if (!componentInfo) throw new Error('Never');
  return [node, componentInfo, genContext] as const;
}

export function getCode(rootCursor: NodePath<JSXElement>) {
  const { code } = generate(rootCursor.node);
  return format(code, { parser: 'babel' });
}
