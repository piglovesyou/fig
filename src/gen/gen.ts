import makeDir from 'make-dir';
import pMap from 'p-map';
import pReduce from 'p-reduce';
import { join } from 'path';
import { requestFile } from '../core/api';
import { FigConfig } from '../core/config';
import { updateLog, updateLogDone } from '../core/print';
import { FigmaFile } from '../types/fig';
import { ComponentInfo, GenContext } from '../types/gen';
import { FigPlugin } from '../types/plugin';
import { writeFile } from '../utils/fs';
import { appendImagesMap, makeExistingImagesMap } from './images-map';
import { makeGenContext, makePaths } from './make-gen-context';
import { processComponent } from './process-component';
import { appendVectorsMap } from './vectors-map';

export async function processHtml(
  componentInfo: ComponentInfo,
  genContext: GenContext
) {
  const { node, name } = componentInfo;
  if (node.type !== 'FRAME')
    throw new Error('Never. node.type should be FRAME');
  const { plugins } = genContext;
  if (!plugins) throw new Error('Never. Plugins should be instantiated.');

  const html = (await pReduce(
    plugins,
    (html, plugin) => {
      return plugin.renderHtml(componentInfo, genContext);
    },
    null as null | ReturnType<FigPlugin<unknown>['renderHtml']>
  ))!;

  await writeFile(join(genContext.htmlFullDir, name + '.html'), html);
}

export async function gen(
  config: FigConfig,
  cwd = process.cwd()
): Promise<void> {
  const { fileKeys, token } = config;
  for (const fileKey of fileKeys) {
    updateLog(`Fetching data of "${fileKey}"..`);
    const figmaFile: FigmaFile = await requestFile(fileKey, token);

    updateLog(`Fetching image resources..`);
    const genContext = await makeGenContext(figmaFile, fileKey, config, cwd);
    const { componentsMap, plugins } = genContext;
    if (!plugins) throw new Error('Never. Plugins should be instantiated.');

    const existingImagesMap = await makeExistingImagesMap(
      genContext.imagesFullDir
    );
    const paths = makePaths(config); // refactor
    await appendImagesMap(
      genContext.imagesMap,
      paths,
      fileKey,
      token,
      existingImagesMap
    );
    await appendVectorsMap(
      paths,
      genContext.vectorsMap,
      genContext.vectorsList,
      fileKey,
      token,
      existingImagesMap
    );

    const components: ComponentInfo[] = [];
    const frames: ComponentInfo[] = [];
    for (const [, componentInfo] of componentsMap) {
      const { node } = componentInfo;
      if (node.type === 'FRAME') frames.push(componentInfo);
      else components.push(componentInfo);
    }

    // Generate components to "./components"
    updateLog(`Generating ${components.length} components..`);
    await pMap(componentsMap, ([, componentInfo]) =>
      processComponent(componentInfo, genContext)
    );

    // Generate components to "./pages"
    updateLog(`Generating ${frames.length} page components..`);
    await pMap(componentsMap, ([, componentInfo]) =>
      processComponent(componentInfo, genContext)
    );

    // Generate html to "./public"
    updateLog(`Generating ${frames.length} HTMLs..`);
    await makeDir(genContext.htmlFullDir);
    await pMap(frames, (componentInfo) => {
      return processHtml(componentInfo, genContext);
    });

    for (const plugin of plugins) await plugin.dispose();

    figmaFile.name;
    updateLog(
      `"${figmaFile.name}" done. ${genContext.imagesMap.size} images, ${components.length} components, ${frames.length} pages and HTMLs are synchronized.`
    );
    updateLogDone();
  }
}
