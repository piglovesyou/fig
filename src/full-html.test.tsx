import { writeFile } from 'fs/promises';
import { join } from 'path';
import React from 'react';
import renderer from 'react-test-renderer';
import * as api from './api';
import { gen } from './gen';
import { readFixtureJson, renderInHtml } from './__tools/fns';

const name = 'full-html';
const outDir = join(process.cwd(), `./__generated__/${name}`);
const fileKey = 'QAIja81RKgYhQnIIJ0h9PJ';

let useCache = false;
// useCache = true; // Comment out to update cache

if (useCache) {
  jest
    .spyOn(api, 'requestFile')
    .mockImplementation(() =>
      readFixtureJson(`./__fixtures/${fileKey}.file.json`)
    );
}

describe('Test full html', () => {
  test(
    'Render final HTML',
    async () => {
      await gen(outDir, fileKey);
      const componentName = 'Home_2$2';
      const { [componentName]: Home } = await import(
        `../__generated__/${name}/components/${componentName}`
      );

      const element = <Home />;

      await writeFile(join(outDir, `index.html`), renderInHtml(<Home />));

      const component = renderer.create(element);
      expect(component.toJSON()).toMatchSnapshot();
    },
    1000 * 60 * 30
  );
});
