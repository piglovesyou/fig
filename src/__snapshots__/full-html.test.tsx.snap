// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test full html Render final HTML 1`] = `
<div
  data-fid="2:2"
  data-fname="Home"
  style={
    Object {
      "backgroundColor": "#ffffff",
      "boxSizing": "border-box",
      "position": "absolute",
    }
  }
>
  <div
    data-fid="28:1"
    data-fname="Arrow 1"
    style={
      Object {
        "boxSizing": "border-box",
        "height": 81,
        "left": 245,
        "position": "absolute",
        "top": 26,
        "width": 70,
        "zIndex": 1,
      }
    }
  >
    <div
      className="vector"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<svg preserveAspectRatio=\\"none\\" width=\\"72\\" height=\\"82\\" viewBox=\\"0 0 72 82\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<path d=\\"M71.0269 0.464673C71.0068 0.18868 70.7667 -0.0187591 70.4907 0.00134421L65.9931 0.328987C65.7171 0.34909 65.5097 0.589126 65.5298 0.865125C65.5499 1.14112 65.7899 1.34856 66.0659 1.32845L70.0638 1.03722L70.355 5.03508C70.3751 5.31108 70.6152 5.51852 70.8912 5.49841C71.1672 5.4783 71.3746 5.23827 71.3545 4.96227L71.0269 0.464673ZM0.758214 82L70.9062 0.828701L70.148 0.173455L8.66095e-07 81.3448L0.758214 82Z\\" fill=\\"black\\"/>
</svg>
",
        }
      }
    />
  </div>
  <div
    data-fid="2:5"
    data-fname="Frame 2"
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "column",
        "gap": 22,
        "height": 122,
        "left": 97,
        "position": "absolute",
        "top": 53,
        "width": 101,
        "zIndex": 0,
      }
    }
  >
    <div
      data-fid="2:3"
      data-fname="Rectangle 1"
      style={
        Object {
          "alignSelf": "inherit",
          "backgroundColor": "#c4c4c4",
          "boxSizing": "border-box",
          "flexGrow": 0,
          "height": 50,
          "position": "relative",
          "width": 101,
          "zIndex": 0,
        }
      }
    />
    <div
      data-fid="2:4"
      data-fname="Rectangle 2"
      style={
        Object {
          "alignSelf": "inherit",
          "backgroundColor": "#c4c4c4",
          "boxSizing": "border-box",
          "flexGrow": 0,
          "height": 50,
          "position": "relative",
          "width": 101,
          "zIndex": 1,
        }
      }
    />
  </div>
  <div
    data-fid="31:5"
    data-fname="Frame 3"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "row",
        "gap": 11,
        "height": 28,
        "left": 220,
        "position": "absolute",
        "top": 215,
        "width": 95,
        "zIndex": 2,
      }
    }
  >
    <div
      data-fid="31:2"
      data-fname="Ellipse 1"
      style={
        Object {
          "alignSelf": "inherit",
          "boxSizing": "border-box",
          "flexGrow": 0,
          "height": 28,
          "position": "relative",
          "width": 28,
          "zIndex": 0,
        }
      }
    >
      <div
        className="vector"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg preserveAspectRatio=\\"none\\" width=\\"28\\" height=\\"28\\" viewBox=\\"0 0 28 28\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<circle cx=\\"14\\" cy=\\"14\\" r=\\"14\\" fill=\\"#C4C4C4\\"/>
</svg>
",
          }
        }
      />
    </div>
    <div
      data-fid="31:3"
      data-fname="aaaaaa"
      style={
        Object {
          "alignSelf": "inherit",
          "boxSizing": "border-box",
          "color": "#000000",
          "flexGrow": 0,
          "fontFamily": "\\"Roboto\\"",
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": 400,
          "height": 14,
          "letterSpacing": "0px",
          "lineHeight": "125%",
          "position": "relative",
          "textAlign": "left",
          "width": 40,
          "zIndex": 1,
        }
      }
    >
      <span
        style={Object {}}
      >
        aaaaaa
      </span>
    </div>
  </div>
  <div
    data-fid="31:5"
    data-fname="Frame 3"
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "row",
        "gap": 11,
        "height": 28,
        "justifyContent": "center",
        "left": 220,
        "position": "absolute",
        "top": 259,
        "width": 95,
        "zIndex": 3,
      }
    }
  >
    <div
      data-fid="31:2"
      data-fname="Ellipse 1"
      style={
        Object {
          "alignSelf": "inherit",
          "boxSizing": "border-box",
          "flexGrow": 0,
          "height": 28,
          "position": "relative",
          "width": 28,
          "zIndex": 0,
        }
      }
    >
      <div
        className="vector"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg preserveAspectRatio=\\"none\\" width=\\"28\\" height=\\"28\\" viewBox=\\"0 0 28 28\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<circle cx=\\"14\\" cy=\\"14\\" r=\\"14\\" fill=\\"#C4C4C4\\"/>
</svg>
",
          }
        }
      />
    </div>
    <div
      data-fid="31:3"
      data-fname="aaaaaa"
      style={
        Object {
          "alignSelf": "inherit",
          "boxSizing": "border-box",
          "color": "#000000",
          "flexGrow": 0,
          "fontFamily": "\\"Roboto\\"",
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": 400,
          "height": 14,
          "letterSpacing": "0px",
          "lineHeight": "125%",
          "position": "relative",
          "textAlign": "left",
          "width": 40,
          "zIndex": 1,
        }
      }
    >
      <span
        style={Object {}}
      >
        aaaaaa
      </span>
    </div>
  </div>
  <div
    data-fid="45:3"
    data-fname="floating"
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "boxSizing": "border-box",
        "display": "flex",
        "flexDirection": "column",
        "gap": 10,
        "height": 145,
        "left": 61,
        "position": "absolute",
        "top": 333,
        "width": 184,
        "zIndex": 4,
      }
    }
  >
    <div
      data-fid="45:4"
      data-fname="Rectangle 7"
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#c4c4c4",
          "boxSizing": "border-box",
          "flexGrow": 1,
          "height": 41.66666793823242,
          "position": "relative",
          "width": 184,
          "zIndex": 0,
        }
      }
    />
    <div
      data-fid="47:0"
      data-fname="Rectangle 8"
      style={
        Object {
          "alignSelf": "inherit",
          "backgroundColor": "#c4c4c4",
          "boxSizing": "border-box",
          "flexGrow": 1,
          "height": 41.66666793823242,
          "position": "relative",
          "width": 92,
          "zIndex": 1,
        }
      }
    />
    <div
      data-fid="45:5"
      data-fname="Frame 4"
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "boxSizing": "border-box",
          "display": "flex",
          "flexDirection": "row",
          "flexGrow": 1,
          "gap": 10,
          "height": 41.66666793823242,
          "position": "relative",
          "width": 184,
          "zIndex": 2,
        }
      }
    >
      <div
        data-fid="45:6"
        data-fname="Rectangle 8"
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#c4c4c4",
            "boxSizing": "border-box",
            "flexGrow": 1,
            "height": 41.66666793823242,
            "position": "relative",
            "width": 128,
            "zIndex": 0,
          }
        }
      />
      <div
        data-fid="45:7"
        data-fname="Rectangle 9"
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#c4c4c4",
            "boxSizing": "border-box",
            "flexGrow": 0,
            "position": "relative",
            "width": 46,
            "zIndex": 1,
          }
        }
      />
    </div>
  </div>
</div>
`;
